wallet: batch erase procedures and improve 'EraseRecords' performance

furszy opened the pull request:
Seeks to optimize and simplify `WalletBatch::EraseRecords`. Currently, this process opens a cursor to iterate over the entire database, searching for records that match the type prefixes, to then call the `WalletBatch::Erase` function for each of the matching records.
This PR rewrites this 40-line manual process into a single line; instead of performing all of those actions manually, we can simply utilize the `ErasePrefix()` functionality. The result is 06216b344dea6ad6c385fda0b37808ff9ae5273b.

Moreover, it expands the test coverage for the `ErasePrefix` functionality and documents the db txn requirement for `BerkeleyBatch::ErasePrefix` .

furszy committed:
wallet: simplify EraseRecords by using 'ErasePrefix'

furszy committed:
wallet: bdb batch 'ErasePrefix', do not create txn internally

Transactions are intended to be started on upper layers rather than
internally by the bdb batch object. This enables us to consolidate
different write operations within a procedure in the same db txn,
improving consistency due to the atomic property of the transaction,
as well as its performance due to the reduction of disk write
operations.

Important Note:
This approach also ensures that the BerkeleyBatch::ErasePrefix
function behaves exactly as the SQLiteBatch::ErasePrefix function,
which does not create a db txn internally.

Furthermore, since the `BerkeleyBatch::ErasePrefix' implementation
erases records one by one (by traversing the db), this change
ensures that the function is always called within an active txn
context. Without this measure, there's a potential risk to consistency;
certain records may be removed while others could persist due to an
internal failure during the procedure.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -2395,15 +2395,37 @@ bool CWallet::DelAddressBook(const CTxDestination& address)
+
+        // Delete purpose and name
+        if (!batch.ErasePurpose(dest) || !batch.EraseName(dest)) {
+            WalletLogPrintf("%s error erasing purpose/name\n", __func__);
```
> Any reason to add `__func__` to this log, when none of the other logs have it here?

Not really. It is just a bad habit. I pulled 9a4e1f5 from #26836 because it is required for this changes. Will change it there.

furszy commented:
Updated per feedback. Thanks Marko.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2396,14 +2397,24 @@ bool CWallet::DelAddressBook(const CTxDestination& address)
+        }
+
+        // Delete purpose and name
+        if (!batch.ErasePurpose(dest) || !batch.EraseName(dest)) {
```
In 2d770f6ae88d9ee6f1775e011f241274059fe275 "wallet: clean redundancies in DelAddressBook"

I'd prefer to have these occur separately so that a distinct error can be logged for each, that way we can discover which operation failed.

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -887,7 +887,7 @@ bool BerkeleyBatch::HasKey(DataStream&& key)
 bool BerkeleyBatch::ErasePrefix(Span<const std::byte> prefix)
 {
-    if (!TxnBegin()) return false;
+    assert(activeTxn); // because this is a read-write operation, the current db config requires a txn context.
```
In 0dba0b6de03c0c79bb09296b913729ec8eb75aca "wallet: bdb batch 'ErasePrefix', do not create txn internally"

Instead of `assert`ing, I would prefer if this did `if (!Assume(activeTxn)) return false` so we didn't cause nodes to crash if a transaction were missing for whatever reason. `Assume` should still catch this for developers.

furszy replied:
> Instead of `assert`ing, I would prefer if this did `if (!Assume(activeTxn)) return false` so we didn't cause nodes to crash if a transaction were missing for whatever reason. `Assume` should still catch this for developers.

That would be risky in terms of consistency. `BerkeleyBatch::ErasePrefix` traverses the entire db, reading entries one by one, erasing the ones that match the key prefix. If we don't execute this function within a txn context, it's possible that certain records are removed while others remain due to an internal failure during the procedure.

Currently, the only function that uses `ErasePrefix()` is `WalletBatch::EraseAddressData`, which is exclusively called from `CWallet::DelAddressBook`, which is modified in the first two commits (thats one of the reasons behind taking those commits from #26836).

achow101 replied:
Sure, that's why I suggest using `Assume`, and `return false` if there is no `activeTxn`. The callers of this should already be checking the return value to know whether the records were actually erased from the database.

furszy replied:
> Sure, that's why I suggest using `Assume`, and `return false` if there is no `activeTxn`. The callers of this should already be checking the return value to know whether the records were actually erased from the database.

ah, I completely skipped the `return false` last night.. sorry. All good, pushed.

achow101 reviewed (commented):
ACK 69c5bb5c3ad3da508541a3ef6c29a0bad21f2fc0

furszy reviewed (commented):
Updated per feedback. Thanks achow!

furszy reviewed (commented):
Updated per feedback. Thanks achow.
[Small diff](https://github.com/bitcoin/bitcoin/compare/0fd7b0dafa980a5dbcee8658d18a83f4eff57085..784034f8a17d46de9eab7de6fdec673a3a3d79d8), only changed `assert` for `Assume`.

furszy committed:
wallet: db, introduce 'RunWithinTxn()' helper function

'RunWithinTxn()' provides a way to execute db operations within a
transactional context. It avoids writing repetitive boilerplate code for
starting and committing the database transaction.

furszy reviewed (commented):
Rebased after #26836 merge. Ready to go.

achow101 commented on src/wallet/test/db_tests.cpp:
```diff
@@ -228,6 +228,39 @@ BOOST_AUTO_TEST_CASE(db_availability_after_write_error)
+        BOOST_CHECK(batch->Write(make_key(value, key), value));
+
+        // Erase the ones with the same prefix and verify result
+        BOOST_CHECK(batch->TxnBegin());
+        BOOST_CHECK(batch->ErasePrefix(DataStream() << key));
+        BOOST_CHECK(batch->TxnCommit());
```
In 523525707742fc620039cd67e8f1437f7f613a01 "wallet: bdb batch 'ErasePrefix', do not create txn internally"

nit: Could use `RunWithinTxn()`.

achow101 commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -447,7 +447,9 @@ BOOST_FIXTURE_TEST_CASE(LoadReceiveRequests, TestingSetup)
             BOOST_CHECK_EQUAL_COLLECTIONS(requests.begin(), requests.end(), std::begin(erequests), std::end(erequests));
             WalletBatch batch{wallet->GetDatabase()};
             BOOST_CHECK(batch.WriteAddressPreviouslySpent(PKHash(), false));
+            BOOST_CHECK(batch.TxnBegin()); // bdb requirement (more info at BerkeleyBatch::ErasePrefix)
             BOOST_CHECK(batch.EraseAddressData(ScriptHash()));
+            BOOST_CHECK(batch.TxnCommit());
```
In 523525707742fc620039cd67e8f1437f7f613a01 "wallet: bdb batch 'ErasePrefix', do not create txn internally"

nit: Could use `RunWithinTxn()`.

furszy replied:
sure

achow101 commented on src/wallet/bdb.cpp:
```diff
@@ -887,7 +887,11 @@ bool BerkeleyBatch::HasKey(DataStream&& key)
 bool BerkeleyBatch::ErasePrefix(Span<const std::byte> prefix)
 {
-    if (!TxnBegin()) return false;
+    // Because this function erases records one by one, ensure that it is executed within a txn context.
+    // Otherwise, consistency is at risk; it's possible that certain records are removed while others
+    // remain due to an internal failure during the procedure.
```
In 523525707742fc620039cd67e8f1437f7f613a01 "wallet: bdb batch 'ErasePrefix', do not create txn internally"

I think it would be relevant to mention that BDB returns errors on `del()` if it is not in a transaction.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29403#discussion_r1483514809

In commit "wallet: bdb batch 'ErasePrefix', do not create txn internally" (96eb527fad0effe082b087c0898843feccc2e213)

The new comment "Additionally, read-write operations performed outside a db txn context are disallowed by current BDB configurations" seems vague to me and I don't think would be comprehensible without reading this review thread. Would suggest just using achow's comment or combining both comments with something like "The Dbc::del() cursor delete call below would fail without an active transaction.".

furszy replied:
> re: [#29403 (comment)](https://github.com/bitcoin/bitcoin/pull/29403#discussion_r1483514809)
> 
> In commit "wallet: bdb batch 'ErasePrefix', do not create txn internally" ([96eb527](https://github.com/bitcoin/bitcoin/commit/96eb527fad0effe082b087c0898843feccc2e213))
> 
> The new comment "Additionally, read-write operations performed outside a db txn context are disallowed by current BDB configurations" seems vague to me and I don't think would be comprehensible without reading this review thread. Would suggest just using achow's comment or combining both comments with something like "The Dbc::del() cursor delete call below would fail without an active transaction.".

I wrote it in this way to (try to) inform readers about current bdb config limitations in case anyone changes this code and doesn't understand why it fails. But, I agree that it is not immediately clear. Will change it for you line.
Hopefully no one else will change this again.

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -1337,6 +1336,12 @@ bool RunWithinTxn(WalletDatabase& database, const std::string& process_desc, con
     return true;
 }
 
+bool RunWithinTxn(WalletDatabase& database, const std::string& process_desc, const std::function<bool(WalletBatch&)>& func)
```
In 757d6516db56e2732ea77624e66025446cb816a4 "wallet: simplify EraseRecords by using 'ErasePrefix'"

This overload feels unnecessary given that there are only 2 places this function is really used.

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -1314,9 +1314,8 @@ DBErrors WalletBatch::ZapSelectTx(std::vector<uint256>& vTxHashIn, std::vector<u
 }
 
-bool RunWithinTxn(WalletDatabase& database, const std::string& process_desc, const std::function<bool(WalletBatch&)>& func)
+static bool RunWithinTxn(WalletBatch& batch, const std::string& process_desc, const std::function<bool(WalletBatch&)>& func)
```
In 757d6516db56e2732ea77624e66025446cb816a4 "wallet: simplify EraseRecords by using 'ErasePrefix'"

This could be done in the commit introducing `RunWithinTxn()` rather than changing it in a later commit.

ryanofsky commented on src/wallet/walletdb.cpp:
```diff
@@ -1314,6 +1314,29 @@ DBErrors WalletBatch::ZapSelectTx(std::vector<uint256>& vTxHashIn, std::vector<u
     return DBErrors::LOAD_OK;
 }
 
+bool RunWithinTxn(WalletDatabase& database, const std::string& process_desc, const std::function<bool(WalletBatch&)>& func)
```
In commit "wallet: db, introduce 'RunWithinTxn()' helper function" (a8a24cb20974cb8eaec8cb438f6b22489719d605)

Maybe use string_view instead of string, since in most cases this is just passed a string literal, so no need to create a full string object

furszy replied:
> Maybe use string_view instead of string, since in most cases this is just passed a string literal, so no need to create a full string object

Sure!

ryanofsky reviewed (approved):
Light code review ACK 757d6516db56e2732ea77624e66025446cb816a4, nice simplification of EraseRecords

furszy reviewed (commented):
Updated per feedback. Thanks both!

achow101 commented:
ACK 8330145ba88b39ddaa9e0bbfc2d1c198da652249

ryanofsky commented on src/wallet/walletdb.cpp:
```diff
@@ -1314,6 +1314,29 @@ DBErrors WalletBatch::ZapSelectTx(std::vector<uint256>& vTxHashIn, std::vector<u
+
+    // Run procedure, errors are logged internally
+    if (!func(batch)) {
+        batch.TxnAbort();
```
In commit "wallet: db, introduce 'RunWithinTxn()' helper function" (b6dcfca8f77f32cb1f41230fb71bab6666252041)

I think it's bad for debugging that this function inconsistently logs errors when an operation fails. For example it seems like there is no error if the new eraserecords function fails? It would be better to add `LogPrint(BCLog::WALLETDB, "Error: %s failed\n", process_desc);` here.

I understand that in the future there may be cases where we may want to abort without logging errors but I think a better way to handle those cases would be to accept `kernel::Interrupted` return values here and use the `kernel::IsInterrupted()` helper. We can move these from `kernel` to `util` without breaking compatibility.

The comment above on line 1325 "errors are logged internally" is also not accurate and impossible to enforce with an anonymous callback, so would suggest dropping that and logging unconditionally here. (Without logging, the most the comment could say would be "errors need to be logged func callback" or "errors should be logged func callback")

furszy replied:
Yeah, using `util::Interrupted` to skip the failure logging in the future sounds great. Error logging added.

ryanofsky commented on src/wallet/walletdb.h:
```diff
@@ -296,6 +296,20 @@ class WalletBatch
+ *
+ * @param database The db connection instance to perform the transaction on.
+ * @param process_desc A description of the process being executed, used for logging purposes in the event of a failure.
+ * @param func The function to be executed within the db txn context. Returns a boolean indicating success or failure.
```
In commit "wallet: db, introduce 'RunWithinTxn()' helper function" (b6dcfca8f77f32cb1f41230fb71bab6666252041)

Would be useful to add this is determines whether to commit or roll back the txn

furszy replied:
Done

ryanofsky reviewed (approved):
Code review ACK 8330145ba88b39ddaa9e0bbfc2d1c198da652249. Looks pretty good, but would suggest cleaning up some things

furszy reviewed (commented):
Updated per feedback. Thanks ryanofsky!. [Small diff](https://github.com/bitcoin/bitcoin/compare/8330145ba88b39ddaa9e0bbfc2d1c198da652249..2095ccdca0cf09073b51b3e9e0d5b9931825824d).
Per request, now `RunWithinTxn` logs an error when the provided function requires to abort the db txn. As well as two comments have been improved for clarity.

achow101 commented:
reACK 77331aa2a198b708372a5c6cdf331faf7e2968ef

josibake commented on src/wallet/walletdb.cpp:
```diff
@@ -1230,9 +1230,8 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
-bool RunWithinTxn(WalletDatabase& database, std::string_view process_desc, const std::function<bool(WalletBatch&)>& func)
+static bool RunWithinTxn(WalletBatch& batch, std::string_view process_desc, const std::function<bool(WalletBatch&)>& func)
 {
-    WalletBatch batch(database);
```
in "wallet: simplify EraseRecords by using 'ErasePrefix'" (https://github.com/bitcoin/bitcoin/pull/29403/commits/77331aa2a198b708372a5c6cdf331faf7e2968ef):

nit: these changes should be in the first commit

josibake commented on src/wallet/walletdb.cpp:
```diff
@@ -1254,6 +1253,12 @@ bool RunWithinTxn(WalletDatabase& database, std::string_view process_desc, const
     return true;
 }
 
+bool RunWithinTxn(WalletDatabase& database, std::string_view process_desc, const std::function<bool(WalletBatch&)>& func)
+{
+    WalletBatch batch(database);
+    return RunWithinTxn(batch, process_desc, func);
+}
+
```
in "wallet: simplify EraseRecords by using 'ErasePrefix'" (https://github.com/bitcoin/bitcoin/pull/29403/commits/77331aa2a198b708372a5c6cdf331faf7e2968ef):

nit: these changes should be in the first commit

josibake reviewed (commented):
code review ACK https://github.com/bitcoin/bitcoin/pull/29403/commits/77331aa2a198b708372a5c6cdf331faf7e2968ef

If you end up retouching, I think it's better to keep all of the `RunWithinTxn` changes in the first commit. Makes it easier to review and avoids introducing the function with one signature and then changing the signature in a later commit.
